File: src/main/java/self/ed/solver/Cell.java
package self.ed.solver;

import java.util.HashSet;
import java.util.Set;

import lombok.Getter;

@Getter
public class Cell {

  private final int row;
  private final int col;
  private final int block;
  private final Set<Integer> candidates;

  public Cell(int row, int col, int block, Set<Integer> candidates) {
    this.row = row;
    this.col = col;
    this.block = block;
    this.candidates = new HashSet<>(candidates);
  }

  public boolean isRelated(Cell cell) {
    return row == cell.row || col == cell.col || block == cell.block;
  }

  public void removeCandidate(Integer value) {
    candidates.remove(value);
  }

  public int countCandidates() {
    return candidates.size();
  }

  public Integer getCandidate() {
    return candidates.iterator().next();
  }
}

File: cpp-sudoku/src/cell.cpp
#include "cell.hpp"

namespace sudoku {

Cell::Cell(int row, int col, int block, const std::unordered_set<int>& candidates)
    : row(row), col(col), block(block), candidates(candidates) {}

bool Cell::isRelated(const Cell& cell) const {
    return row == cell.row || col == cell.col || block == cell.block;
}

std::optional<int> Cell::getCandidate() const {
    if (candidates.empty()) {
        return std::nullopt;
    }
    return *candidates.begin();
}

} // namespace sudoku
